sx.el

```diff
@@ -259,6 +259,23 @@ whenever BODY evaluates to nil."
        :filter (lambda (&optional _)
                  (when (progn ,@body) ,def)))))
 
+(defmacro sx--create-comparator (name doc compare-func get-func)
+  "Define a new comparator called NAME with documentation DOC.
+COMPARE-FUNC is a function that takes the return value of
+GET-FUNC and performs the actual comparison."
+  (declare (indent 1) (doc-string 2))
+  (let ((gf (intern (format " %S--get-prop-function" name)))
+        (cf (intern (format " %S--compare-function"  name))))
+    ;; Leading space to hide from completion systems
+    `(progn
+       ;; In using `defalias', the macro supports both function
+       ;; symbols and lambda expressions.
+       (defalias ',gf ,get-func)
+       (defalias ',cf ,compare-func)
+       (defun ,name (a b)
+         ,doc
+         (,cf (,gf a) (,gf b))))))
```

Comment:
Kind of. I think we're not really evaluating it twice, we're byte-compiling it twice. When we write `(something a b)`, `a` and `b` are evaluated, but `something` isn't. If `something` is a lambda list, it'll be byte-compiled, but that happens outside the macro (after the macro is expanded) and I don't know if it's even good practice to try to mess with that.

OTOH, I think `cl-flet` may solve all of this. Note, I'm saying `flet` not `letf`. :-) `letf` won't work because it's dynamic.

What `cl-flet` does is look into its body and manually replace all calls using a specified function name with the definition you provide. This essentially makes it lexical, instead of dynamic. But personally, I think it's overkill.